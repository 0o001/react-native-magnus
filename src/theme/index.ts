export { ThemeContext } from './themeContext';
export { withTheme } from './withTheme';
export { ThemeProvider, ThemeProviderProps } from './themeProvider.component';

export {
  ThemeType,
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
} from './type';
