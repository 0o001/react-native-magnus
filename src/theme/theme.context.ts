import * as React from 'react';

import { ThemeType } from './type';

const defaultValue: ThemeType = {};

export interface ThemeContextType {
  theme: ThemeType;
  setTheme?: (theme?: ThemeType) => void;
}

export const THEME_CONTEXT_DEFAULT_VALUE = {
  theme: defaultValue,
};

export const ThemeContext: React.Context<ThemeContextType> = React.createContext(
  THEME_CONTEXT_DEFAULT_VALUE
);
