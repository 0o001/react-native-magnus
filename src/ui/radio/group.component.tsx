import * as React from "react";

import { Div } from "../div/div.component";
import { DivProps } from "../div/div.type";

interface ButtonProps extends DivProps {
  onChange?: any;
  value?: any;
  children: React.ReactElement[] | React.ReactElement;
}

const RadioGroup: React.FunctionComponent<ButtonProps> = props => {
  const { children, onChange, value, ...rest } = props;

  /**
   * clones the children and add currentValue
   */
  const renderChildren = () => {
    return React.Children.map(children, (child: React.ReactElement) => {
      return React.cloneElement(child, {
        currentValue: value,
        onChange
      });
    });
  };

  return <Div {...rest}>{renderChildren()}</Div>;
};

export { RadioGroup };
