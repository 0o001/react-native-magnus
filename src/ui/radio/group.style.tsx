import { StyleSheet } from 'react-native';

import {
  createSpacingStyles,
  createBorderRadiusStyles,
  createBorderWidthStyles,
  createBorderColorStyles,
} from '../../theme/theme.service';

/**
 * computed style
 *
 * @param theme
 * @param props
 */
export const getStyle = (theme: any, props: any) => {
  const computedStyle: any = {};

  computedStyle.container = {
    ...createBorderWidthStyles(props),
    ...createSpacingStyles(props, theme.spacing),
    ...createBorderColorStyles(props, theme.colors),
    ...createBorderRadiusStyles(props, theme.borderRadius),
  };

  // merging style props to computed style
  if (props.style) {
    computedStyle.div = {
      ...computedStyle.div,
      ...props.style,
    };
  }

  return StyleSheet.create(computedStyle);
};
