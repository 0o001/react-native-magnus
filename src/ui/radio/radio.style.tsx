import { StyleSheet } from 'react-native';

import {
  getThemeProperty,
  createSpacingStyles,
  createBorderRadiusStyles,
  createBorderWidthStyles,
  createBorderColorStyles,
} from '../../theme/theme.service';

/**
 * computed style
 *
 * @param theme
 * @param props
 */
export const getStyle = (theme: any, props: any) => {
  const computedStyle: any = {};

  computedStyle.div = {
    borderColor: getThemeProperty(
      theme.colors,
      props.borderColor,
      'transparent'
    ),
  };

  computedStyle.container = {
    flexDirection: 'row',
    alignItems: 'center',
    ...createBorderWidthStyles(props),
    ...createSpacingStyles(props, theme),
    ...createBorderColorStyles(props, theme.colors),
    ...createBorderRadiusStyles(props, theme.borderRadius),
  };

  computedStyle.text = {
    fontSize: getThemeProperty(theme.fontSize, props.fontSize, 16),
  };

  computedStyle.highlightContainer = {
    backgroundColor: getThemeProperty(theme.colors, 'gray100', '#e1e1e1'),
    height: getThemeProperty(theme.fontSize, props.size, 16) + 6,
    width: getThemeProperty(theme.fontSize, props.size, 16) + 6,
    alignItems: 'center',
    justifyContent: 'center',
    top: -3,
    left: -3,
    borderRadius: getThemeProperty(theme.fontSize, props.size, 16),
    zIndex: 1,
    position: 'absolute',
  };

  // merging style props to computed style
  if (props.style) {
    computedStyle.div = {
      ...computedStyle.div,
      ...props.style,
    };
  }

  return StyleSheet.create(computedStyle);
};
