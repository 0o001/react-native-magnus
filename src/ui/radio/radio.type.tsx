import {
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
} from '../../theme';

export interface RadioProps
  extends BorderPropsType,
    SpacingPropsType,
    RoundedPropsType {
  value?: any;
  style?: any;
  color?: string;
  onChange?: any;
  size?: string;
  checked?: boolean;
  loading?: boolean;
  currentValue?: any;
  disabled?: boolean;
  activeColor?: string;
  inactiveColor?: string;
}
