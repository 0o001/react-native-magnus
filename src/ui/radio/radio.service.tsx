import { getThemeProperty } from '../../theme/theme.service';

/**
 * get icon name to be used based on checked state
 *
 * @param checked
 */
export const getIconName = (checked: boolean = false) => {
  switch (true) {
    case checked:
      return 'radio-button-checked';
    default:
      return 'radio-button-unchecked';
  }
};

/**
 * get icon color based on state
 *
 * @param checked
 * @param disabled
 */
export const getIconColor = (
  checked: any,
  disabled: any,
  activeColor: any,
  inactiveColor: any,
  theme: any
) => {
  switch (true) {
    case disabled:
      return getThemeProperty(theme.colors, inactiveColor);
    case checked:
      return getThemeProperty(theme.colors, activeColor);
    default:
      return getThemeProperty(theme.colors, inactiveColor);
  }
};
