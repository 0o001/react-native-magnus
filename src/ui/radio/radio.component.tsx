import * as React from 'react';
import { useContext, useState } from 'react';
import {
  View as RNView,
  ActivityIndicator,
  TouchableWithoutFeedback as RNButton,
} from 'react-native';
import { getStyle } from './radio.style';
import { RadioProps } from './radio.type';
import { ThemeContext } from '../../theme';
import { Icon } from '../icon/icon.component';
import { getIconName, getIconColor } from './radio.service';
import { getThemeProperty } from '../../theme/theme.service';

const Radio: React.FunctionComponent<RadioProps> = (props) => {
  const {
    m,
    mt,
    mr,
    mb,
    ml,
    p,
    pr,
    pt,
    pb,
    pl,
    size,
    value,
    style,
    loading,
    disabled,
    onChange,
    children,
    activeColor,
    currentValue,
    inactiveColor,
    ...rest
  } = props;
  const { theme } = useContext(ThemeContext);
  const [focussed, setFocussed] = useState(false);
  const computedStyle = getStyle(theme, props);

  /**
   * on press checkbox
   */
  const onPress = () => {
    const { disabled, onChange, value } = props;
    if (disabled) {
      return;
    }

    onChange(value);
  };

  /**
   * sets focussed to true
   *
   * @param event
   */
  const onPressIn = () => {
    setFocussed(true);
  };

  /**
   * sets focussed to true
   *
   * @param event
   */
  const onPressOut = () => {
    setFocussed(false);
  };

  /**
   * get icon
   * shows activity indication if loading state is true
   */
  const getIcon = () => {
    return loading ? (
      <ActivityIndicator
        size={getThemeProperty(theme.fontSize, size, 16)}
        color={iconColor}
      />
    ) : (
      <Icon
        name={iconName}
        color={iconColor}
        fontFamily="MaterialIcons"
        fontSize={size}
        w={getThemeProperty(theme.fontSize, size, 16)}
        h={getThemeProperty(theme.fontSize, size, 16)}
        style={{ zIndex: 2, position: 'relative' }}
      />
    );
  };

  const checked = currentValue === value;
  const iconName = getIconName(checked);
  const iconColor = getIconColor(
    checked,
    disabled,
    activeColor,
    inactiveColor,
    theme
  );
  const icon = getIcon();

  return (
    <RNButton
      {...rest}
      style={computedStyle.button}
      onPress={onPress}
      onPressIn={onPressIn}
      onPressOut={onPressOut}
    >
      <RNView style={computedStyle.container}>
        <RNView>
          {focussed && <RNView style={computedStyle.highlightContainer} />}
          {icon}
        </RNView>
        {children}
      </RNView>
    </RNButton>
  );
};

Radio.defaultProps = {
  loading: false,
  disabled: false,
  size: 'text600',
  value: null,
  activeColor: 'blue600',
  inactiveColor: 'gray900',
};

export { Radio };
