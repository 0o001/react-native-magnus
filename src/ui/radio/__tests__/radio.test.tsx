import React from "react";
import renderer from "react-test-renderer";

import { Radio } from "../radio.component";
import { ThemeProvider } from "../../../theme";

describe("Radio Component", () => {
  it("should render without issues", () => {
    const instance = renderer.create(
      <ThemeProvider>
        <Radio />
      </ThemeProvider>
    );

    expect(instance.toJSON()).toMatchSnapshot();
  });
});
