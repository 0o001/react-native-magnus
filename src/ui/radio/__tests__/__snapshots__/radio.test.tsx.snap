// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Component should render without issues 1`] = `
<View
  accessible={true}
  clickable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "color": "#000",
          "fontSize": 19,
          "height": 19,
          "justifyContent": "center",
          "position": "relative",
          "width": 19,
          "zIndex": 2,
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#262626",
              "fontSize": 19,
            },
            undefined,
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        î ¶
      </Text>
    </View>
  </View>
</View>
`;
